torch/csrc/distributed/rpc/python_rpc_handler.h

```diff
@@ -39,6 +39,21 @@ class PYBIND11_EXPORT PythonRpcHandler {
   // Deserialize a string into a py::object
   py::object deserialize(const SerializedPyObj& serializedObj);
 
+  // Explicitly clean up py::objects to avoid segment faults when
+  // py::objects with CPython are cleaned up later at program exit
+  // See similar issues reported https://github.com/pybind/pybind11/issues/1598
+  // and https://github.com/pybind/pybind11/issues/1493
+  // Our local tests also caught this segment faults if py::objects are cleaned
+  // up at program exit. The explaination is: CPython cleans up most critical
+  // utitlies before cleaning up PythonRpcHandler singleton, so when
+  // PythonRpcHandler signleton cleans up py::objects and call dec_ref(), it
+  // will crash.
+  // The solution is to clean up py::objects earlier when Rpc agent join().
+  // Be note that py::objects can not be cleaned up when Rpc agent is destroyed
+  // as well, as Rpc agent is global variable and it will have same issue as
+  // PythonRpcHandler.
```

Comment:
Just make sure I understand the above statement. Are you saying that we cannot put `cleanup` in `~ProcessGroupAgent` due to the same issue?

Reply:
yes, I tried it before, it has same issue, because at that point, Cpython may clean up some critical utilities and not able to clean up the py::objects correctly sometimes